Refactor serialization getter/setters into torch.utils.serialization.config

mikaylagawarecki opened the pull request:
Consolidate 
- get/set_default_load_endianness
- get/set_default_mmap_options
- get/set_crc32_options

into one global dynamo-style config + allow global setting of mmap. The existing APIs are not removed and will get/set from the config (as they can't be removed for BC)

In #143459 I add the local (argument style) config



Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #143342
* __->__ #143324

mikaylagawarecki committed:
Refactor serialization getter/setters into torch.serialization.config

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.serialization.config"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.serialization.config"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.serialization.config"




[ghstack-poisoned]

mikaylagawarecki commented on torch/utils/_config_module.py:
```diff
@@ -161,6 +161,11 @@ def visit(
                 or (hasattr(value, "__module__") and value.__module__ == "typing")
                 # Handle from torch.utils._config_module import Config
                 or (isinstance(value, type) and issubclass(value, _Config))
+                # Handle from torch.serialization import LoadEndianness
+                or (
+                    hasattr(value, "__module__")
+                    and value.__module__ == "torch.serialization"
+                )
```
cc @ezyang is it okay to do something like this?

tldr: this visitor will take any instance of type within the file as a sub config and check that its __module__ is the config module

In my case I [import an enum from `torch.serialization`](https://github.com/pytorch/pytorch/pull/143324/files#diff-284f52965247307a704d9a8ea0de055ccb76ae521a90f1c1deb5a98c5989077dR4) that is the type that one of the configs should be, but is not intended to be a subconfig.


This enum goes into the `elif isinstance(value, type):` case, but it's not intended for this to be parsed as a subconfig. Is it okay to prevent this by hardcode it like what was done for `Config` above?

ezyang replied:
If you want to bodge it this way you can but we should figure out a better way that doesn't involve everyone adding their own special cases here. @c00w

albanD replied:
It feels like changing the variable name in the import to start with double underscore would make it being ignored per the first rule above? `__LoadEndianess`

mikaylagawarecki replied:
@albanD I did try this, using `__LoadEndianness` seems to cause some issues

Traceback (most recent call last):
  File "/data/users/mg1998/pytorch/test/test_serialization.py", line 26, in <module>
    from torch.utils.serialization import config as serialization_config
  File "/data/users/mg1998/pytorch/torch/utils/serialization/__init__.py", line 1, in <module>
    from . import config
  File "/data/users/mg1998/pytorch/torch/utils/serialization/config.py", line 8, in <module>
    class load:
  File "/data/users/mg1998/pytorch/torch/utils/serialization/config.py", line 10, in load
    endianness: _Optional[__LoadEndianess] = None
NameError: name '_load__LoadEndianess' is not defined. Did you mean: '__LoadEndianess'?

c00w replied:
I think here we can fix this by making this conditional on typechecking. But we probably should rewrite this code to ignore import statements. It's not 100% clear to me how to do that, but I'm sure it's possible.

albanD replied:
The usual way to filter import statement would actually be to check if the `__module__` field does not match the name of the module you're current config-ifying. It is gameable by changing this field, but at least it will be correct for everything defined in the file

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.utils.serialization.config"







[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.utils.serialization.config"







[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.utils.serialization.config"

Consolidate 
- get/set_default_load_endianness
- get/set_default_mmap_options
- get/set_crc32_options

into one config + allow global setting of mmap






[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.utils.serialization.config"

Consolidate 
- get/set_default_load_endianness
- get/set_default_mmap_options
- get/set_crc32_options

into one config + allow global setting of mmap






[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.utils.serialization.config"

Consolidate 
- get/set_default_load_endianness
- get/set_default_mmap_options
- get/set_crc32_options

into one config + allow global setting of mmap






[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.utils.serialization.config"

Consolidate 
- get/set_default_load_endianness
- get/set_default_mmap_options
- get/set_crc32_options

into one config + allow global setting of mmap

The existing APIs are not removed (as they can't be for BC)






[ghstack-poisoned]

albanD commented on docs/source/notes/serialization.rst:
```diff
@@ -487,3 +487,29 @@ The following utility functions are related to serialization:
+
+``torch.utils.serialization.config.save`` contains options that control the behavior of ``torch.save``.
+
+  * ``compute_crc32``: whether to compute and write crc32 for each storage (Default : ``True``).
```
```suggestion
  * ``compute_crc32``: whether to compute and write the zip file checksum (Default : ``True``).
```
Would this be more accurate? Isn't this skip writing the global checksum, not just the storage one?

albanD commented on torch/utils/serialization/config.py:
```diff
@@ -0,0 +1,20 @@
+import sys
+from typing import Optional as _Optional
+
+from torch.serialization import LoadEndianness as _LoadEndianess
```
To avoid the circular import that forces you to do in-function import in the other file, should this enum be defined here and just imported in the the serialization.py file?

albanD replied:
Ok, let's keep it as is then.

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.utils.serialization.config"

Consolidate 
- get/set_default_load_endianness
- get/set_default_mmap_options
- get/set_crc32_options

into one global dynamo-style config + allow global setting of mmap. The existing APIs are not removed and will get/set from the config (as they can't be removed for BC)

In #143459 I add the local (argument style) config






[ghstack-poisoned]

c00w commented on torch/utils/serialization/_config.py:
```diff
@@ -0,0 +1,26 @@
+
+class load:
+    mmap: bool = False
+    endianness: Optional[LoadEndianness] = None
```
Weird question - Can we do

``
Optional["LoadEndianness"]
```

and above

```
if TYPECHECKING:
  from torch.serialization import LoadEndianness
```

mikaylagawarecki replied:
This seems to work :)

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.utils.serialization.config"

Consolidate 
- get/set_default_load_endianness
- get/set_default_mmap_options
- get/set_crc32_options

into one global dynamo-style config + allow global setting of mmap. The existing APIs are not removed and will get/set from the config (as they can't be removed for BC)

In #143459 I add the local (argument style) config






[ghstack-poisoned]

albanD reviewed (approved):
SGTM if others are happy with the config part!

mikaylagawarecki commented:
Good catch, I [accidentally squashed the config update into the previous PR](https://github.com/pytorch/pytorch/pull/143324/files#diff-030970443ab6ae89bef63cf45fecfaec09d906ac677a736c5129878275681859R20), let me fix that

mikaylagawarecki committed:
Update on "Refactor serialization getter/setters into torch.utils.serialization.config"

Consolidate 
- get/set_default_load_endianness
- get/set_default_mmap_options
- get/set_crc32_options

into one global dynamo-style config + allow global setting of mmap. The existing APIs are not removed and will get/set from the config (as they can't be removed for BC)

In #143459 I add the local (argument style) config






[ghstack-poisoned]
